src/node/transaction.cpp

```diff
@@ -52,27 +53,28 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
             CallFunctionInValidationInterfaceQueue([&promise] {
                 promise.set_value();
             });
+            callback_set = true;
         }
     } else if (fHaveChain) {
         return TransactionError::ALREADY_IN_CHAIN;
-    } else {
-        // Make sure we don't block forever if re-sending
-        // a transaction already in mempool.
-        promise.set_value();
     }
 
     } // cs_main
 
-    promise.get_future().wait();
+    if (callback_set) {
+        promise.get_future().wait();
+    }
 
     if (!g_connman) {
```

Comment:
Should this also be within the `if (relay)` block? Returning a `TransactionError::P2P_DISABLED` error when the caller hasn't asked for the transaction to be relayed seems wrong.